### PR TITLE
Update package building scripts and images to use Conan

### PR DESCRIPTION
**Pull request by legleux**
## High Level Overview of Change

Lots of changes needed to update the docker containers we use to build DEB and RPM packages.

### Context of Change

rippled's dependencies are handled by Conan now so the package building process needs to be updated.

bump version to rc3 
test on Fedora 37
rm docker user mapping  on container builds (doesn't work with conan) 
rm dpkg dev pkg (was broken anyway)
rm dependency builds (dependencies are handled by Conan now)

Supersedes https://github.com/XRPLF/rippled/pull/4362

### Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Refactor (non-breaking change that only restructures code)
- [ ] Tests (You added tests for code that already exists, or your new feature included in this PR)
- [ ] Documentation Updates
- [ ] Release

### CODE REVIEW

**Comment by legleux** [comment]
@thejohnfreeman I'm still unsure about the removal from `RippledRelease.cmake` of

```
# Early return if the `containers` directory is missing,
# e.g. when we are building a Conan package.
if(NOT EXISTS containers)
  return()
endif()
```

File: Builds/containers/ubuntu-builder/ubuntu_setup.sh [hunk 1]
```diff
@@ -34,129 +31,99 @@ apt-get -y --fix-missing install \
...
-    --slave /usr/bin/llvm-nm llvm-nm /usr/bin/llvm-nm-8
+    /usr/bin/clang clang /usr/bin/clang-14 20 \
+    --slave /usr/bin/clang++ clang++ /usr/bin/clang++-14 \
+    --slave /usr/bin/llvm-profdata llvm-profdata /usr/bin/llvm-profdata-14 \
+    --slave /usr/bin/asan-symbolize asan-symbolize /usr/bin/asan_symbolize-14 \
+    --slave /usr/bin/llvm-symbolizer llvm-symbolizer /usr/bin/llvm-symbolizer-14 \
+    --slave /usr/bin/clang-format clang-format /usr/bin/clang-format-14 \
+    --slave /usr/bin/llvm-ar llvm-ar /usr/bin/llvm-ar-14 \
+    --slave /usr/bin/llvm-cov llvm-cov /usr/bin/llvm-cov-14 \
+    --slave /usr/bin/llvm-nm llvm-nm /usr/bin/llvm-nm-14
```
Comment (thejohnfreeman) [inline Builds/containers/ubuntu-builder/ubuntu_setup.sh:124]: If these two blocks are the same, can we do this?

```bash
for v in 12 14; do
  ... $v ...
done
```

File: Builds/containers/ubuntu-builder/ubuntu_setup.sh [hunk 2]
```diff
@@ -185,5 +152,9 @@ if [ "${CI_USE}" = true ] ; then
     update-alternatives --auto clang
 fi
 
-apt-get -y autoremove
+pip install "conan==1.58" && \
```
Comment (thejohnfreeman) [inline Builds/containers/ubuntu-builder/ubuntu_setup.sh:155]: Consider changing to `conan<2`.

File: Builds/containers/ubuntu-builder/Dockerfile [hunk 3]
```diff
@@ -1,36 +1,28 @@
...
-RUN /tmp/build_deps.sh
-ENV PLANTUML_JAR="/opt/plantuml/plantuml.jar"
-ENV BOOST_ROOT="/opt/local/boost/_INSTALLED_"
 ENV OPENSSL_ROOT="/opt/local/openssl"
 
 # prep files for package building
-RUN mkdir -m 777 -p /opt/rippled_bld/pkg/debian
-RUN update-alternatives --set gcc /usr/bin/gcc-8
+RUN update-alternatives --set gcc /usr/bin/gcc-11
+RUN mkdir -m 777 -p /opt/rippled_bld/pkg/
```
Comment (thejohnfreeman) [inline Builds/containers/ubuntu-builder/Dockerfile:24]: I'd like to see almost every `RUN` statement in this Dockerfile, including these, moved to `ubuntu_setup.sh`. Best to have all the commands in one place, that another person can execute themselves without having to build a Docker image. My personal workflow when developing these install scripts is to log in interactively to a Docker container and copy bits of the script over via tmux; checking the output; and adding, removing, and changing bits as needed.

File: Builds/containers/packaging/rpm/rippled.spec [hunk 4]
```diff
@@ -36,30 +36,76 @@ History server for XRP Ledger
...
-cmake .. -G Ninja -DCMAKE_INSTALL_PREFIX=%{_prefix} -DCMAKE_BUILD_TYPE=Release -Dunity=OFF -Dstatic=true -DCMAKE_VERBOSE_MAKEFILE=OFF -Dvalidator_keys=ON
+conan install .. --output-folder . --build missing --settings build_type=Release
+cmake -G Ninja \
+     -DCMAKE_TOOLCHAIN_FILE:FILEPATH=build/generators/conan_toolchain.cmake \
+     -DCMAKE_INSTALL_PREFIX=%{_prefix} \
+     -DCMAKE_BUILD_TYPE=Release \
+     -Dunity=OFF \
+     -Dstatic=ON \
+     -Dvalidator_keys=ON \
+     -DCMAKE_BUILD_TYPE=Release \
```
Comment (thejohnfreeman) [inline Builds/containers/packaging/rpm/rippled.spec:66]: Duplicate of line 62.

File: Builds/containers/packaging/rpm/rippled.spec [hunk 5]
```diff
@@ -36,30 +36,76 @@ History server for XRP Ledger
...
+conan install .. --output-folder . --build missing --settings build_type=Release -s compiler.cppstd=17 --options reporting=True
+
+cmake -G Ninja \
+     -DCMAKE_TOOLCHAIN_FILE:FILEPATH=build/generators/conan_toolchain.cmake \
+     -DCMAKE_INSTALL_PREFIX=%{_prefix} \
+     -DCMAKE_BUILD_TYPE=Release \
+     -Dunity=OFF \
+     -Dstatic=ON \
+     -Dvalidator_keys=ON \
+     -DCMAKE_BUILD_TYPE=Release \
```
Comment (thejohnfreeman) [inline Builds/containers/packaging/rpm/rippled.spec:85]: Duplicate of line 81.

File: Builds/containers/packaging/rpm/rippled.spec [hunk 6]
```diff
@@ -36,30 +36,76 @@ History server for XRP Ledger
...
 %build
+rm -rf ~/.conan/profiles/default
+
+cp /opt/libcstd/libstdc++.so.6.0.22 /usr/lib64
+cp /opt/libcstd/libstdc++.so.6.0.22 /lib64
+ln -sf /usr/lib64/libstdc++.so.6.0.22 /usr/lib64/libstdc++.so.6
+ln -sf /lib64/libstdc++.so.6.0.22 /usr/lib64/libstdc++.so.6
+
+source /opt/rh/rh-python38/enable
+pip install "conan==1.58"
```
Comment (thejohnfreeman) [inline Builds/containers/packaging/rpm/rippled.spec:47]: Consider `conan<2`.

File: Builds/containers/packaging/rpm/rippled.spec [hunk 7]
```diff
@@ -36,30 +36,76 @@ History server for XRP Ledger
 %setup -c -n rippled
 
 %build
+rm -rf ~/.conan/profiles/default
+
+cp /opt/libcstd/libstdc++.so.6.0.22 /usr/lib64
+cp /opt/libcstd/libstdc++.so.6.0.22 /lib64
+ln -sf /usr/lib64/libstdc++.so.6.0.22 /usr/lib64/libstdc++.so.6
+ln -sf /lib64/libstdc++.so.6.0.22 /usr/lib64/libstdc++.so.6
```
Comment (thejohnfreeman) [inline Builds/containers/packaging/rpm/rippled.spec:44]: Are these still necessary?

File: Builds/containers/packaging/dpkg/debian/rules [hunk 8]
```diff
@@ -16,31 +16,36 @@ override_dh_systemd_start:
...
 	env
-	rm -rf bld && mkdir -p bld/rippled
-	cd bld/rippled && \
-	cmake ../.. -G Ninja \
-		-DCMAKE_INSTALL_PREFIX=/opt/ripple \
+	rm -rf bld
+
+	conan export external/snappy snappy/1.1.9@
+
+	conan install . -if bld/rippled  --build missing --settings build_type=Release
```
Comment (thejohnfreeman) [inline Builds/containers/packaging/dpkg/debian/rules:23]: I prefer long options in these scripts, especially if their short form lends itself to being misunderstood.

```suggestion
	conan install . --install-folder bld/rippled  --build missing --settings build_type=Release
```

File: Builds/containers/packaging/dpkg/debian/rules [hunk 9]
```diff
@@ -16,31 +16,36 @@ override_dh_systemd_start:
...
+		-DCMAKE_VERBOSE_MAKEFILE=ON \
 		-Dvalidator_keys=ON \
-		-Dunity=OFF \
-		-DCMAKE_VERBOSE_MAKEFILE=OFF
-
+		-B bld/rippled
 
-	cmake -S . \
-		-B bld/rippled-reporting \
+	conan install . -if bld/rippled-reporting --build missing --settings build_type=Release --options reporting=True -s compiler.cppstd=17
```
Comment (thejohnfreeman) [inline Builds/containers/packaging/dpkg/debian/rules:34]:
```suggestion
	conan install . --install-folder bld/rippled-reporting --build missing --settings build_type=Release --options reporting=True --options cassandra-cpp-driver:use_atomic=std
```

C++17 might have worked, but I don't like that solution. Did you try the other option I suggested?

File: Builds/containers/centos-builder/Dockerfile [hunk 10]
```diff
@@ -1,43 +1,26 @@
...
 RUN /tmp/centos_setup.sh
 
-RUN /tmp/install_cmake.sh 3.16.1 /opt/local/cmake-3.16
+RUN /tmp/install_cmake.sh 3.16.3 /opt/local/cmake-3.16
 RUN ln -s /opt/local/cmake-3.16 /opt/local/cmake
 ENV PATH="/opt/local/cmake/bin:$PATH"
-# also install min supported cmake for testing
-RUN if [ "${CI_USE}" = true ] ; then /tmp/install_cmake.sh 3.9.0 /opt/local/cmake-3.9; fi
+# TODO: Install latest CMake for testing
+RUN if [ "${CI_USE}" = true ] ; then /tmp/install_cmake.sh 3.16.3 /opt/local/cmake-3.16; fi
```
Comment (thejohnfreeman) [inline Builds/containers/centos-builder/Dockerfile:17]: See my comments on the Ubuntu Dockerfile for my thoughts on moving all these `RUN` commands to the install script and greatly simplifying / trivializing the Dockerfile.

File: Builds/CMake/RippledRelease.cmake [hunk 11]
```diff
@@ -2,12 +2,6 @@
    package/container targets - (optional)
 #]===================================================================]
 
-# Early return if the `containers` directory is missing,
-# e.g. when we are building a Conan package.
-if(NOT EXISTS containers)
-  return()
-endif()
-
```
Comment (thejohnfreeman) [inline Builds/CMake/RippledRelease.cmake:10]: If this worked for you, then it is safe to remove. I think I added it when I was debugging Conan packaging in a personal branch. We're not there yet, though.

**Comment by intelliot** [comment]
I guess this might be a flaky test. I kicked it off again.
```
1> ripple.app.LedgerData
0> #299 failed: apply: Got temINVALID (The transaction is ill-formed.); Expected tesSUCCESS (The transaction was applied. Only final in a validated ledger.)
0> {
   "Account" : "rHb9CJAWyB4rj91VRWn96DkukG4bwdtyTh",
   "Amount" : "1000000010",
   "Destination" : "rJP5kd48NUvKuEMUUNdByiQCauJVYURBz8",
   "Fee" : "10",
   "Flags" : 2147483648,
   "Sequence" : 51,
   "SigningPubKey" : "0330E7FC9D56BB25D6893BA3F317AE5BCF33B3291BD63DB32654A313222F7FD020",
   "TransactionType" : "Payment",
   "TxnSignature" : "304502210092D9DA00AAE0404640A8697028B3E36BA45FCA7[500](https://github.com/XRPLF/rippled/actions/runs/4266933705/jobs/7428013711#step:19:501)D475FAC8F6B97B01B1D80902206B51127825AD3AE0ED3F7CE16CC665533571BAF49B852B4B1E69DBD552BFA4E0"
}

0> #300 failed: unhandled exception: unexpected missing account root
0> ripple.app.LedgerReplayer ProofPath
```

File: Builds/containers/shared/install_cmake.sh [hunk 12]
```diff
@@ -21,14 +21,17 @@ if [[ "${installed}" != "" && ${installed} =~ ${cm_maj}.${cm_min}.${cm_rel} ]] ;
     echo "cmake already installed: ${installed}"
     exit
 fi
-
-pkgname="cmake-${cm_maj}.${cm_min}.${cm_rel}-$(uname)-x86_64.tar.gz"
+# From CMake 20+ "Linux" is lowercase so using `uname` won't create be the correct path
```
[the remark this answers is not in the document]
Comment (thejohnfreeman) [reply Builds/containers/shared/install_cmake.sh:24]: Ok. Let's just hard-code "Linux" in a future change then.

**Review by thejohnfreeman — APPROVED** [review]

File: Builds/containers/ubuntu-builder/ubuntu_setup.sh [hunk 13]
```diff
@@ -34,156 +31,74 @@ apt-get -y --fix-missing install \
...
-    /usr/bin/gcc gcc /usr/bin/gcc-7 40 \
-    --slave /usr/bin/g++ g++ /usr/bin/g++-7 \
-    --slave /usr/bin/gcc-ar gcc-ar /usr/bin/gcc-ar-7 \
-    --slave /usr/bin/gcc-nm gcc-nm /usr/bin/gcc-nm-7 \
-    --slave /usr/bin/gcc-ranlib gcc-ranlib /usr/bin/gcc-ranlib-7 \
-    --slave /usr/bin/gcov gcov /usr/bin/gcov-7 \
-    --slave /usr/bin/gcov-tool gcov-tool /usr/bin/gcov-dump-7 \
-    --slave /usr/bin/gcov-dump gcov-dump /usr/bin/gcov-tool-7
+if [[ ${VERSION_ID} =~ ^20\. ]] ; then
+apt-get install -y \
```
Comment (sgramkumar) [inline Builds/containers/ubuntu-builder/ubuntu_setup.sh:45]: nit: indentation missing

File: Builds/containers/ubuntu-builder/ubuntu_setup.sh [hunk 14]
```diff
@@ -34,156 +31,74 @@ apt-get -y --fix-missing install \
...
-    --slave /usr/bin/clang-format clang-format /usr/bin/clang-format-8 \
-    --slave /usr/bin/llvm-ar llvm-ar /usr/bin/llvm-ar-8 \
-    --slave /usr/bin/llvm-cov llvm-cov /usr/bin/llvm-cov-8 \
-    --slave /usr/bin/llvm-nm llvm-nm /usr/bin/llvm-nm-8
-update-alternatives --auto clang
 
-if [ "${CI_USE}" = true ] ; then
+apt-get -y update
+if [[ ${VERSION_ID} =~ ^20\. ]] ; then
+for v in 12 14; do
```
Comment (sgramkumar) [inline Builds/containers/ubuntu-builder/ubuntu_setup.sh:81]: nit: indentation missing

**Review by sgramkumar — COMMENTED** [review]
LGTM. Just a minor indentation missing

**Review by HowardHinnant — APPROVED** [review]
I started with a empty `~/.conan/data`, and it installed, generated, built and tested without errors for me on macOS, using the directions in BUILD.md.

**Review by sgramkumar — APPROVED** [review]

**Comment by intelliot** [comment]
@legleux to confirm this is ready to merge, you can put the `Passed` label on it (or comment saying it's ready)

**Comment by legleux** [comment]
How about that, I can add labels?
Let's merge it.

**Pull request merged**